h/browser/chrome/lib/browser-action.js

```diff
@@ -36,23 +36,88 @@
       }
     };
 
+    /**
+     * Set the "title" (tooltip) of the browser action _if_ no badge is
+     * currently displayed on the browser action.
+     * @param {integer} tabId The id of the tab to set the badge title for.
+     * @param {string} title The value to set the title to.
+     */
+    function setTitleIfNoBadge(tabId, title) {
+      chromeBrowserAction.getBadgeText({tabId: tabId}, function(text) {
+        if (!text) {
+          chromeBrowserAction.setTitle({tabId: tabId, title: title});
+        }
+      });
+    }
+
+    /**
+     * Set the "title" (tooltip) and badge text of the browser action.
+     * @param {integer} tabId The id of the tab to set the badge title for.
+     * @param {string} title The text to show in the tooltip.
+     * @param {string} badgeText The text to show on the badge.
+     */
+    function setTitleAndBadgeText(tabId, title, badgeText) {
+      chromeBrowserAction.setTitle({tabId: tabId, title: title});
+      chromeBrowserAction.setBadgeText({tabId: tabId, text: badgeText});
+    }
+
     /* Sets the active browser action appearance for the provided tab id. */
-    this.activate = function (tabId) {
+    this.activate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.ACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is active')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is active'));
     };
 
     /* Sets the inactive browser action appearance for the provided tab id. */
-    this.deactivate = function (tabId) {
+    this.deactivate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is inactive')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is inactive'));
     };
 
     /* Sets the errored browser action appearance for the provided tab id. */
-    this.error = function (tabId) {
+    this.error = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis has failed to load')});
-      chromeBrowserAction.setBadgeText({tabId: tabId, text: '!'});
+      setTitleAndBadgeText(tabId,  _('Hypothesis has failed to load'), '!');
+    };
+
+    /**
+     * Show the number of annotations of the current page in the badge.
+     * @method
+     * @param {integer} tabId The id of the current tab.
+     * @param {string} tabUrl The URL of the current tab.
+     * @param {string} serviceUrl The URL of the Hypothesis API.
+     */
+    this.updateBadge = function(tabId, tabUrl, serviceUrl) {
+      // Fetch the number of annotations of the current page from the server,
+      // and display it as a badge on the browser action button.
+      var xhr = new XMLHttpRequest();
+      xhr.onload = function() {
+        var total = 0;
+        total = JSON.parse(this.response).total;
```

Comment:
What happens if for some reason the response contains data that isn't valid JSON, or the JSON isn't correctly formatted (e.g. has no `total` field)? Does the background.html crash? Does the extension continue to work?

If it's possible that bad data will crash the extension, this function needs to deal with that possibility.